### PR TITLE
LPS-189352 | LPS-189353 Perform values comparison by dbType

### PR DESCRIPTION
**Pull request by p-albuquerque**
Related issues:
https://issues.liferay.com/browse/LPS-189352;
https://issues.liferay.com/browse/LPS-189353.

### CODE REVIEW

File: modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java [hunk 1]
```diff
@@ -280,12 +283,38 @@ private static void _validateNewValue(
...
+				GetterUtil.getBoolean(value),
+				GetterUtil.getBoolean(existingValue))) {
 
-			throw new ObjectFieldReadOnlyException(
-				"Object field " + objectFieldName + " is read only");
+			return;
 		}
+
+		try {
+			BigDecimal bigDecimal1 = new BigDecimal(existingValue.toString());
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java:299]: Can we test if `value` is an instance of BigDecimal here? I'm afraid of always hit the catch condition for all other types of fields.

File: modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java [hunk 2]
```diff
@@ -280,12 +283,38 @@ private static void _validateNewValue(
...
-		if (!((Validator.isNull(existingValue) && Validator.isNull(value)) ||
-			  Objects.equals(existingValue, value))) {
+		if (Validator.isNull(existingValue) && Validator.isNull(value)) {
+			return;
+		}
+
+		if ((value instanceof Boolean) &&
+			Objects.equals(
+				GetterUtil.getBoolean(value),
+				GetterUtil.getBoolean(existingValue))) {
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java:293]: I don't know if it's a possible scenario but if `existingValue` was not a instance of boolean, GetterUtil.getBoolean will return false value.

**Review by guilhermedcamacho — CHANGES_REQUESTED** [review]

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

**Comment by p-albuquerque** [comment]
ci:stop

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

File: modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java [hunk 3]
```diff
@@ -277,15 +282,75 @@ else if (Objects.equals(
...
+			}
 		}
+		else if (Objects.equals(dbType, ObjectFieldConstants.DB_TYPE_DATE) ||
+				 Objects.equals(
+					 dbType, ObjectFieldConstants.DB_TYPE_DATE_TIME)) {
+
+			String existingDateValue = String.valueOf(existingValue);
+
+			if (Objects.equals(dbType, ObjectFieldConstants.DB_TYPE_DATE)) {
+				existingDateValue = existingDateValue.substring(0, 10);
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java:319]: Hey @p-albuquerque , why do we need to get substring for date values?

File: modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java [hunk 4]
```diff
@@ -277,15 +282,75 @@ else if (Objects.equals(
...
+			}
+			else if (Objects.equals(
+						dbType, ObjectFieldConstants.DB_TYPE_DATE_TIME)) {
+
+				existingDateValue = existingDateValue.substring(0, 16);
+				existingDateValue = existingDateValue.replaceAll(
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java:325]: Why do we need to get substring for date-time values? Also, why do we need to replace characters by space?
I'm afraid of these manipulations goes wrong.
Did you check/try other approaches used in Portal, like [DateUtil](https://github.com/liferay/liferay-portal/blob/master/portal-kernel/src/com/liferay/portal/kernel/util/DateUtil.java) or [CompareDatesFunction](https://github.com/liferay/liferay-portal/blob/master/modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-evaluator-impl/src/main/java/com/liferay/dynamic/data/mapping/form/evaluator/internal/function/CompareDatesFunction.java)

**Review by guilhermedcamacho — CHANGES_REQUESTED** [review]

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

**Comment by p-albuquerque** [comment]
ci:test:sf

**Comment by p-albuquerque** [comment]
ci:test:relevant

File: modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java [hunk 5]
```diff
@@ -282,17 +314,40 @@ else if (Objects.equals(
...
 		throws PortalException {
 
 		if (Validator.isNull(existingValue) && Validator.isNull(value)) {
 			return;
 		}
 
-		if (Objects.equals(dbType, ObjectFieldConstants.DB_TYPE_BLOB) ||
-			Objects.equals(dbType, ObjectFieldConstants.DB_TYPE_CLOB) ||
-			Objects.equals(dbType, ObjectFieldConstants.DB_TYPE_STRING)) {
+		if (Objects.equals(
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/field/util/ObjectFieldUtil.java:324]: Just to let you know that you can also use `ObjectFieldImpl#compareBusinessType` for this.

This change is not mandatory, but if you want, you can change it here or send a follow-up PR.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java [hunk 6]
```diff
@@ -1330,6 +1283,53 @@ public void testEnableAccountEntryRestrictedForNondefaultStorageType()
 		_objectDefinitionLocalService.deleteObjectDefinition(objectDefinition);
 	}
 
+	@Test
+	public void testEnableAccountRestricted() throws Exception {
```
Comment (guilhermedcamacho) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectDefinitionLocalServiceTest.java:1287]: Did you change this class/method intentionally?

**Review by guilhermedcamacho — APPROVED** [review]
LGTM :tada:

**Comment by p-albuquerque** [comment]
ci:forward:force

**Pull request closed without merging**